dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -447,6 +490,37 @@ static void AddResponseMessage(ChatCompletionsOptions chatOptions, ChatHistory c
         }
     }
 
+    private void HandleStopBehavior(ToolFilterContext context, ChatCompletionsOptions chatOptions, ref bool autoInvoke)
+    {
+        switch (context.StopBehavior)
+        {
+            case ToolFilterStopBehavior.StopAutoInvoke:
+                autoInvoke = false;
+                break;
+            case ToolFilterStopBehavior.StopTools:
+                chatOptions.ToolChoice = ChatCompletionsToolChoice.None;
+                break;
+            case ToolFilterStopBehavior.Cancel:
+                throw new OperationCanceledException();
+        }
+    }
+
+    private void UpdateChatHistory(ChatHistory chatHistory, ChatCompletionsOptions options, OpenAIPromptExecutionSettings executionSettings)
```

Comment:
This is being called every time a filter is called right? Is that to handle the case of the filter handler changes the history in some way and we need to sync it? Do we do any updates to the history to clean up if the tool call is canceled?

Reply:
> This is being called every time a filter is called right? Is that to handle the case of the filter handler changes the history in some way and we need to sync it? 

Yes and yes. If the filter changes the chat history, we would want that updated chat history to be included in the next request to the model. This helper function updates the chat history stored in `chatOptions` (which is what is sent to the model) to match what is in the `chatHistory` object.

> Do we do any updates to the history to clean up if the tool call is canceled?

If the tool call is cancelled *before* invocation, instead of a response message, we add an error message to the chat history. If the tool call is cancelled *after* invocation, I just have it keeping the regular tool response message in the chat history, but bailing out of future tool calls. (I am not necessarily opposed to a different behavior here, but this made the most sense to me.) In both these cases, `AddResponseMessage` handles updating both the `chatHistory` and `chatOptions` objects.